node/actors/bft/src/testonly/twins/mod.rs

```diff
@@ -10,125 +10,45 @@
+pub trait HasKey {
+    /// Validator public key.
+    type Key: PartialEq + Ord;
 
-/// Recursive partition generator.
-struct Partitioner<'a, T> {
-    items: Partition<'a, T>,
-    num_items: usize,
-    num_partitions: usize,
-    // All collected complete partitionings
-    output: Vec<Partitioning<'a, T>>,
-    // Partially complete partitioning currently being built
-    acc: Partitioning<'a, T>,
+    /// The validator key of this node, used to identify who is leading a round.
+    ///
+    /// The twin and its original both have the same key.
+    fn key(&self) -> &Self::Key;
 }
 
-impl<'a, T> Partitioner<'a, T> {
-    /// Generate all possible partitioning.
-    fn generate(items: Partition<'a, T>, num_partitions: usize) -> Vec<Partitioning<'a, T>> {
-        if num_partitions == 0 || items.len() < num_partitions {
-            // Impossible to partition.
-            Vec::new()
-        } else if num_partitions == items.len() {
-            // Each items stands alone.
-            vec![items.into_iter().map(|i| vec![i]).collect()]
-        } else if num_partitions == 1 {
-            // All items are in a single partition.
-            vec![vec![items]]
-        } else {
-            // Create empty partitions for each slot.
-            let acc = (0..num_partitions)
-                .into_iter()
-                .map(|_| Vec::new())
-                .collect();
+impl<T: HasKey> HasKey for &T {
+    type Key = T::Key;
 
-            let mut ps = Self {
-                num_items: items.len(),
-                num_partitions,
-                items,
-                output: Vec::new(),
-                acc,
-            };
-            // Generate all combinations
-            ps.go(0, 0);
-            // Take the results
-            ps.output
-        }
+    fn key(&self) -> &Self::Key {
+        (*self).key()
     }
+}
 
-    /// Recursively generate partitions.
-    ///
-    /// The algorithm goes item-by-item and has two indexes depending on what has happened
-    /// above it: the minimum and the maximum partition that it can (or has to) insert into.
-    ///
-    /// Take the table in [partitions] as an example with 4 items and 3 partitions.
-    /// * `A` is the first item, so it can only go into `P1` - `P2` can't be used
-    /// while `P1` is empty, otherwise we'd be generating redundant combinations. After that,
-    /// * `B` can go into minimally into `P1` because there are 2 more items after it, which
-    /// is enough to fill all remaining partitions; or it can go into `P2`, because `P1` is filled.
-    /// * `C` depends on what we did with `B`: if `B` is in `P1` then `C` has to minimally go into
-    /// `P2` to make sure no partition will be left empty at the end; if `B` is in `P2` then `C`
-    /// can go either in `P1`, `P2` or `P3`
+/// Abstraction for nodes which can have a twin.
+pub trait Twin: HasKey
+where
+    Self: PartialEq,
+{
```

Comment:
nit: I trust that you have some need to make this abstract, but afaic having concrete types was also ok. Using traits has its own limitations so use your judgment.

Reply:
It's not that I absolutely needed traits to be abstract, I just like to develop in a way where I can ignore stuff like what exactly is going to be used for this or that type (like a `bn245` or whatever key is used) when it's completely irrelevant to this case. If we work with abstractions, then there can't be any tight coupling either, and the minimum expectations can be expressed. Code is also easier to extract, should we want to. 

The `Twin` itself is just an idea for how to deal with these things. It turns out it's not really needed, there could be a constructor which puts the job of creating twins on the caller, but I thought it can't hurt to have it, to make it easier to understand what's going on.